File: libraries/broker/bindings/python/zeek.cpp

#include <utility>
#include <string>

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
#include <pybind11/pybind11.h>
#pragma GCC diagnostic pop

#include "broker/zeek.hh"
#include "broker/data.hh"

namespace py = pybind11;
using namespace pybind11::literals;

void init_zeek(py::module& m) {
  py::class_<broker::zeek::Message>(m, "Message")
    .def("as_data",
         static_cast<const broker::data& (broker::zeek::Message::*)() const>
         (&broker::zeek::Message::as_data));

  py::class_<broker::zeek::Event, broker::zeek::Message>(m, "Event")
    .def(py::init([](broker::data data) {
       return broker::zeek::Event(std::move(data));
       }))
    .def(py::init([](std::string name, broker::data args) {
       return broker::zeek::Event(std::move(name), std::move(caf::get<broker::vector>(args)));
       }))
    .def("name",
          static_cast<const std::string& (broker::zeek::Event::*)() const>
          (&broker::zeek::Event::name))
    .def("args",
         static_cast<const broker::vector& (broker::zeek::Event::*)() const>
         (&broker::zeek::Event::args));
}

File: libraries/broker/bindings/python/broker/zeek.py

try:
    from . import _broker
except ImportError:
    import _broker

import broker

class Event(_broker.zeek.Event):
    def __init__(self, *args):
        if len(args) == 1 and not isinstance(args[0], str):
            # Parse raw broker message as event.
            _broker.zeek.Event.__init__(self, broker.Data.from_py(args[0]))
        else:
            # (name, arg1, arg2, ...)
            _broker.zeek.Event.__init__(self, args[0], broker.Data.from_py(args[1:]))

    def args(self):
        return [broker.Data.to_py(a) for a in _broker.zeek.Event.args(self)]
